src/native/directory_traversal.c

```diff
@@ -0,0 +1,85 @@
+/**
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+#include "crt.h"
+#include "java_class_ids.h"
+#include "retry_utils.h"
+#include <aws/common/file.h>
+#include <aws/common/string.h>
+#include <jni.h>
+
+/* on 32-bit platforms, casting pointers to longs throws a warning we don't need */
+#if UINTPTR_MAX == 0xffffffff
+#    if defined(_MSC_VER)
+#        pragma warning(push)
+#        pragma warning(disable : 4305) /* 'type cast': truncation from 'jlong' to 'jni_tls_ctx_options *' */
+#        pragma warning(disable : 4221)
+#    else
+#        pragma GCC diagnostic push
+#        pragma GCC diagnostic ignored "-Wpointer-to-int-cast"
+#        pragma GCC diagnostic ignored "-Wint-to-pointer-cast"
+#    endif
+#endif
+
+struct directory_traversal_callback_ctx {
+    JNIEnv *env;
+    jobject handler;
+};
+
+static bool s_on_directory_entry(const struct aws_directory_entry *entry, void *user_data) {
+
+    struct directory_traversal_callback_ctx *ctx = user_data;
+    JNIEnv *env = ctx->env;
+
+    jstring path = aws_jni_string_from_cursor(env, &entry->path);
+    jstring relativePath = aws_jni_string_from_cursor(env, &entry->relative_path);
+
+    jboolean callback_result = (*env)->CallBooleanMethod(
+        env,
+        ctx->handler,
+        directory_traversal_handler_properties.on_directory_entry_method_id,
+        path,
+        relativePath,
+        (jboolean)entry->file_type & AWS_FILE_TYPE_DIRECTORY,
+        (jboolean)entry->file_type & AWS_FILE_TYPE_SYM_LINK,
+        (jboolean)entry->file_type & AWS_FILE_TYPE_FILE,
+        (jlong)entry->file_size);
+
+    return (bool)callback_result;
```

Comment:
kinda sucks that an exception gets swallowed here, and just stops iteration like nothing went wrong. might be a good idea to modify aws-c-common's implementation so the callback returns an int and we can alert when something goes wrong.

we'd need to do something else to stop, either a special "safe" error like AWS_ERROR_STOP or a separate bool out-param somewhere

ALSO: use aws_jni_check_and_clear_exception() to check whether an exception happened, and properly clear it

Reply:
If an exception is thrown by the user callback, the traversal is cancelled. Cancelling the traversal either via the user callback returning `false` or throwing, results in `DirectoryTraversal.traverse()` throwing an exception to notify user about incomplete results. This behavior is checked by the unit tests.

Added the missing `aws_jni_check_and_clear_exception()`. Curiously it was working before because the return value was `false` if an exception was thrown from inside the callback.

Reply:
Oh hey, this is a synchronous callback, we can actually preserve the java exception!

instead of calling `aws_jni_check_and_clear_exception(env)`, which clears the exception, just check if the exception occurred via `(*env)->ExceptionCheck(env)`

and if an exception occurred, then stop the iteration by returning false.

and then in the `Java_software_amazon_awssdk_crt_io_DirectoryTraversal_crtTraverse()`, the JNI exception will still be set. I think that when that function returns, since the JNI exception is still set, you'll see the exception propagate!

Does that make sense? typing quick before the end of the day

Reply:
Good idea, thanks! :-)